### ISSUE TITLE
GIT and Missing tree / MissingObjectException

### ISSUE THREAD

**Issue by HagarJNode** [spring-cloud/spring-cloud-config#522]
Running the Spring Cloud Config server it sometimes gives an exception and display the error page. We reboot the server to make it recover, but as we get it in production it would be nice to find the reason. It's valid calls that fails, rebooting will make the calls respond with correct config data.

```
2016-10-08 21:35:30.083 ERROR 7128 --- [nio-8080-exec-7] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is java.lang.IllegalStateException: Cannot load environment] with root cause

org.eclipse.jgit.errors.MissingObjectException: Missing tree b0fb112b1cc6e7e490bf049e1c037ab1b8d9bcb3
    at org.eclipse.jgit.internal.storage.file.WindowCursor.open(WindowCursor.java:149) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar!/:3.5.3.201412180710-r]
    at org.eclipse.jgit.treewalk.CanonicalTreeParser.reset(CanonicalTreeParser.java:202) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar!/:3.5.3.201412180710-r]
    at org.eclipse.jgit.treewalk.CanonicalTreeParser.createSubtreeIterator0(CanonicalTreeParser.java:236) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar!/:3.5.3.201412180710-r]
    at org.eclipse.jgit.treewalk.CanonicalTreeParser.createSubtreeIterator(CanonicalTreeParser.java:214) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar!/:3.5.3.201412180710-r]
    at org.eclipse.jgit.treewalk.CanonicalTreeParser.createSubtreeIterator(CanonicalTreeParser.java:60) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar!/:3.5.3.201412180710-r]
    at org.eclipse.jgit.treewalk.TreeWalk.enterSubtree(TreeWalk.java:912) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar!/:3.5.3.201412180710-r]
    at org.eclipse.jgit.dircache.DirCacheCheckout.preScanTwoTrees(DirCacheCheckout.java:259) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar!/:3.5.3.201412180710-r]
    at org.eclipse.jgit.dircache.DirCacheCheckout.doCheckout(DirCacheCheckout.java:410) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar!/:3.5.3.201412180710-r]
    at org.eclipse.jgit.dircache.DirCacheCheckout.checkout(DirCacheCheckout.java:397) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar!/:3.5.3.201412180710-r]
    at org.eclipse.jgit.api.CheckoutCommand.call(CheckoutCommand.java:261) ~[org.eclipse.jgit-3.5.3.201412180710-r.jar!/:3.5.3.201412180710-r]
    at org.springframework.cloud.config.server.environment.JGitEnvironmentRepository.checkout(JGitEnvironmentRepository.java:204) ~[spring-cloud-config-server-1.1.3.RELEASE.jar!/:1.1.3.RELEASE]
    at org.springframework.cloud.config.server.environment.JGitEnvironmentRepository.refresh(JGitEnvironmentRepository.java:148) ~[spring-cloud-config-server-1.1.3.RELEASE.jar!/:1.1.3.RELEASE]
    at org.springframework.cloud.config.server.environment.JGitEnvironmentRepository.getLocations(JGitEnvironmentRepository.java:121) ~[spring-cloud-config-server-1.1.3.RELEASE.jar!/:1.1.3.RELEASE]
    at org.springframework.cloud.config.server.environment.MultipleJGitEnvironmentRepository.getLocations(MultipleJGitEnvironmentRepository.java:111) ~[spring-cloud-config-server-1.1.3.RELEASE.jar!/:1.1.3.RELEASE]
    at org.springframework.cloud.config.server.environment.AbstractScmEnvironmentRepository.findOne(AbstractScmEnvironmentRepository.java:40) ~[spring-cloud-config-server-1.1.3.RELEASE.jar!/:1.1.3.RELEASE]
    at org.springframework.cloud.config.server.environment.MultipleJGitEnvironmentRepository.findOne(MultipleJGitEnvironmentRepository.java:146) ~[spring-cloud-config-server-1.1.3.RELEASE.jar!/:1.1.3.RELEASE]
    at org.springframework.cloud.config.server.environment.EnvironmentEncryptorEnvironmentRepository.findOne(EnvironmentEncryptorEnvironmentRepository.java:53) ~[spring-cloud-config-server-1.1.3.RELEASE.jar!/:1.1.3.RELEASE]
    at org.springframework.cloud.config.server.environment.EnvironmentController.labelled(EnvironmentController.java:112) ~[spring-cloud-config-server-1.1.3.RELEASE.jar!/:1.1.3.RELEASE]
.
.
.
```

**Comment by spencergibb** [comment]
What version are you using? Googling leads me to this https://www.eclipse.org/forums/index.php/t/205022/

**Comment by HagarJNode** [comment]
We are using the 1.1.3.RELEASE, but it also happened with 1.1.2.RELEASE... Note that there was no changes to the repository, so it's not a matter of merging as in the Eclipse issue. We do frequently fetch different branches/labels from our systems tho, maybe some kind of hint? The repository is placed at GitLab, but I don't see that as an issue.

**Comment by HagarJNode** [comment]
You're using the same org.eclipse.jgit version, 3.5.3.201412180710-r, on both 1.1.3.RELEASE and 1.2.0.RELEASE... I think there is a chance that we might see the problem in the newest release too.

**Comment by HagarJNode** [comment]
Just a following up... I updated to 1.2, but the problem continued... Finally set up a loadbalanced setup in AWS with two nodes, each in different availability zone, made the problem happen less often and are now very rare. Second, if errors and health check fails, the instance is replaced by automatic. My guess is, that two requests to the cloud config server on different labels/branches, on the "same" time, could corrupt the local git checkout.

**Comment by kinpa0a** [comment]
I have a similar issue. But I'm not sure whether it's the same.

In my case, I use a git as a backend and a commit number as a label.
When Spring Cloud Config server is running and new commit is created, that exception occurs.
If I make a request with the commit number to get a config data, it fails calling JGitEnvironmentRepository#checkout.

As I guess, it wants to checkout to "label" but it can't find the label.
That's because new commit is not fetched or pulled yet from remote that time.

We tried to fetch new commit from remote before calling JGitEnvironmentRepository#refresh and it worked. But I'm curious about this solution. Is the "detour" right to solve that issue?

**Comment by rlynch** [comment]
Try upgrading to 1.2.2 as this was hopefully fixed by #538.

**Comment by kinpa0a** [comment]
@rlynch It works for me :-) Thanks!
